src/manager.cpp

```diff
@@ -0,0 +1,218 @@
+    }
+    catch (const std::exception& e)
+    {
+        log<level::ERR>("Failure to close an obmc session.",
+                        entry("SESSION-ID=%s", sessionId.c_str()),
+                        entry("ERROR=%s", e.what()));
+        throw InternalFailure();
+    }
+}
+
+std::size_t SessionManager::removeAll(const std::string& userName)
+{
+    const auto count =
+        std::erase_if(sessionItems, [userName](const auto& item) {
+            auto const [id, session] = item;
+            return session->getOwner() == userName;
+        });
+
+    return count;
+}
+
+std::size_t
+    SessionManager::removeAllByRemoteAddress(const std::string& remoteAddress)
+{
+    const auto count =
+        std::erase_if(sessionItems, [remoteAddress](const auto& item) {
+            auto const [id, session] = item;
+            return session->remoteIPAddr() == remoteAddress;
+        });
+
+    return count;
+}
+
+std::size_t SessionManager::removeAll()
+{
+    size_t handledSessions = sessionItems.size();
+    sessionItems.clear();
+    return handledSessions;
+}
+
+SessionIdentifier SessionManager::generateSessionId() const
+{
+    auto time = std::chrono::high_resolution_clock::now();
+    std::size_t timeHash =
+        std::hash<int64_t>{}(time.time_since_epoch().count());
+    std::size_t serviceNameHash = std::hash<std::string>{}(serviceName);
+
+    std::size_t result = timeHash ^ (serviceNameHash << 1);
+    if (result == invalidSessionId)
+    {
+        // The hash-collision guard. The Session ID == invalidSessionId is
+        // reserved and can't be provided as a valid ID.
+        return generateSessionId();
+    }
+
+    return result;
+}
+
+const std::string
+    SessionManager::getSessionObjectPath(SessionIdentifier sessionId) const
+{
+    return getSessionManagerObjectPath() + "/" + hexSessionId(sessionId);
+}
+
+const std::string SessionManager::getSessionManagerObjectPath() const
+{
+    return std::string(sessionManagerObjectPath);
+}
+
+const std::string SessionManager::hexSessionId(SessionIdentifier sessionId)
+{
+    std::stringstream stream;
+    stream << std::setfill('0') << std::setw(sizeof(sessionId) * 2) << std::hex
+           << sessionId;
+    return stream.str();
+}
+
+SessionIdentifier
+    SessionManager::parseSessionId(const std::string& hexSessionId)
+{
+    return std::stoull(hexSessionId, nullptr, 16);
+}
+
+void SessionManager::checkSessionOwnerAlive(const boost::system::error_code&)
+{
+    for (auto it = sessionItems.begin(); it != sessionItems.end();)
+    {
+        if (it->second && !fs::exists(it->second->getProcPath()))
+        {
+            const auto sessionId = hexSessionId(it->first);
+            log<level::DEBUG>(
+                "Found unreachable service",
+                entry("SESSION=%s", sessionId.c_str()),
+                entry("SVC_PROC=%s", it->second->getProcPath().c_str()));
+            this->close(sessionId);
+            it = sessionItems.erase(it);
```

Comment:
`it` is already invalid as `this->close(sessionId)` erases specified item.
I believe you should either return an iterator from `close()` or not call `close()` here at all.
As I can see the `close()` doesn't do any important things, but do another search on the map.
So it may be safely skipped here.

Reply:
My bad. I will fix this

Reply:
fixed